File: apps/web/src/lib/api/inventory-ownership.ts
import { apiFetch } from '@/lib/api/client';

export const OWNERSHIP_TYPES = [
  'legal_owner',
  'beneficial_owner',
  'economic_owner',
  'joint_owner',
  'trustee',
  'nominee',
  'manager',
  'other',
] as const;

export type OwnershipType = (typeof OWNERSHIP_TYPES)[number];

export const TRANSFER_TYPES = [
  'initial_ownership',
  'full_transfer',
  'partial_transfer',
  'percentage_change',
  'owner_addition',
  'owner_removal',
  'ownership_type_change',
  'correction',
  'other',
] as const;

export type TransferType = (typeof TRANSFER_TYPES)[number];

export const TRANSFER_PARTY_ROLES = [
  'outgoing_owner',
  'incoming_owner',
  'continuing_owner',
] as const;

export type TransferPartyRole = (typeof TRANSFER_PARTY_ROLES)[number];

export const TRANSFER_REQUEST_STATUSES = [
  'draft',
  'submitted',
  'under_review',
  'approved',
  'rejected',
  'revision_requested',
  'withdrawn',
  'applied',
  'stale',
] as const;

export type TransferRequestStatus = (typeof TRANSFER_REQUEST_STATUSES)[number];

export interface OwnershipRecord {
  id: string;
  inventory_asset_id: string;
  party_id: string;
  party_name: string | null;
  party_type: string | null;
  ownership_type: OwnershipType;
  ownership_percentage: string;
  effective_from: string;
  effective_to: string | null;
  status: string;
  acquisition_method: string;
  transfer_reason: string | null;
  related_document_id: string | null;
  related_transaction_id: string | null;
  source: string;
  notes: string | null;
  created_by_user_id: string | null;
  approved_by_user_id: string | null;
  ownership_transfer_request_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface TransferParty {
  id: string;
  party_id: string;
  party_name: string | null;
  party_type: string | null;
  ownership_type: OwnershipType;
  previous_percentage: string | null;
  proposed_percentage: string;
  role: TransferPartyRole;
  notes: string | null;
  created_at: string;
}

export interface OwnershipTransferRequest {
  id: string;
  inventory_asset_id: string;
  asset_display_id: string | null;
  project_name: string | null;
  transfer_type: TransferType;
  effective_date: string;
  reason: string;
  supporting_document_id: string | null;
  related_transaction_id: string | null;
  requested_by_user_id: string;
  assigned_approver_user_id: string | null;
  status: TransferRequestStatus;
  reviewed_at: string | null;
  approved_at: string | null;
  rejected_at: string | null;
  decision_notes: string | null;
  parties: TransferParty[];
  actions: string[];
  proposed_legal_total: string | null;
  current_legal_total: string | null;
  created_at: string;
  updated_at: string;
}

export interface AssetOwnershipDetail {
  asset_id: string;
  current_legal: OwnershipRecord[];
  current_beneficial: OwnershipRecord[];
  current_economic: OwnershipRecord[];
  current_other: OwnershipRecord[];
  pending_requests: OwnershipTransferRequest[];
  scheduled_transfers: OwnershipTransferRequest[];
  history: OwnershipRecord[];
  legal_total: string;
}

export interface OwnershipSummary {
  primary_legal_owner: string | null;
  owner_count: number;
  ownership_status: string;
}

export interface TransferPartyInput {
  party_id: string;
  ownership_type: OwnershipType;
  previous_percentage?: string | null;
  proposed_percentage: string;
  role: TransferPartyRole;
  notes?: string | null;
}

export interface TransferRequestInput {
  inventory_asset_id: string;
  transfer_type: TransferType;
  effective_date: string;
  reason: string;
  parties: TransferPartyInput[];
  supporting_document_id?: string | null;
  related_transaction_id?: string | null;
  assigned_approver_user_id?: string | null;
  submit?: boolean;
}

export function fetchAssetOwnershipDetail(assetId: string): Promise<AssetOwnershipDetail> {
  return apiFetch(`/inventory/ownership/by-asset/${assetId}/detail`);
}

export function fetchOwnershipSummary(assetId: string): Promise<OwnershipSummary> {
  return apiFetch(`/inventory/ownership/summary/by-asset/${assetId}`);
}

export function fetchPendingOwnershipApprovals(): Promise<OwnershipTransferRequest[]> {
  return apiFetch('/inventory/ownership-transfers/pending-approvals');
}

export function createOwnershipTransfer(input: TransferRequestInput): Promise<OwnershipTransferRequest> {
  return apiFetch('/inventory/ownership-transfers', {
    method: 'POST',
    body: JSON.stringify(input),
  });
}

export function submitOwnershipTransfer(requestId: string): Promise<OwnershipTransferRequest> {
  return apiFetch(`/inventory/ownership-transfers/${requestId}/submit`, { method: 'POST' });
}

export function approveOwnershipTransfer(
  requestId: string,
  comments?: string,
): Promise<OwnershipTransferRequest> {
  return apiFetch(`/inventory/ownership-transfers/${requestId}/approve`, {
    method: 'POST',
    body: JSON.stringify({ comments: comments ?? null }),
  });
}

export function rejectOwnershipTransfer(
  requestId: string,
  decisionNotes: string,
): Promise<OwnershipTransferRequest> {
  return apiFetch(`/inventory/ownership-transfers/${requestId}/reject`, {
    method: 'POST',
    body: JSON.stringify({ decision_notes: decisionNotes }),
  });
}

export function requestOwnershipRevision(
  requestId: string,
  decisionNotes: string,
): Promise<OwnershipTransferRequest> {
  return apiFetch(`/inventory/ownership-transfers/${requestId}/request-revision`, {
    method: 'POST',
    body: JSON.stringify({ decision_notes: decisionNotes }),
  });
}

export function formatPercentage(value: string | number | null | undefined, locale: string): string {
  if (value === null || value === undefined || value === '') return '—';
  const num = typeof value === 'string' ? Number.parseFloat(value) : value;
  if (Number.isNaN(num)) return '—';
  return new Intl.NumberFormat(locale, { maximumFractionDigits: 2 }).format(num) + '%';
}

File: apps/api/src/investhome_api/schemas/inventory_ownership.py
"""Inventory ownership API schemas."""

from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from investhome_api.models.inventory import (
    AcquisitionMethod,
    OwnershipApprovalDecision,
    OwnershipRecordStatus,
    OwnershipSource,
    OwnershipType,
    TransferPartyRole,
    TransferRequestStatus,
    TransferType,
)


class TransferPartyInput(BaseModel):
    party_id: UUID
    ownership_type: OwnershipType
    previous_percentage: Decimal | None = None
    proposed_percentage: Decimal = Field(..., ge=Decimal("0.0001"), le=Decimal("100"))
    role: TransferPartyRole
    notes: str | None = None


class OwnershipTransferRequestCreate(BaseModel):
    inventory_asset_id: UUID
    transfer_type: TransferType
    effective_date: date
    reason: str = Field(min_length=1)
    parties: list[TransferPartyInput] = Field(min_length=1)
    supporting_document_id: UUID | None = None
    related_transaction_id: UUID | None = None
    assigned_approver_user_id: UUID | None = None
    submit: bool = False


class OwnershipTransferRequestUpdate(BaseModel):
    effective_date: date | None = None
    reason: str | None = Field(default=None, min_length=1)
    parties: list[TransferPartyInput] | None = None
    supporting_document_id: UUID | None = None
    related_transaction_id: UUID | None = None
    assigned_approver_user_id: UUID | None = None


class TransferPartyResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    party_id: UUID
    party_name: str | None = None
    party_type: str | None = None
    ownership_type: OwnershipType
    previous_percentage: Decimal | None
    proposed_percentage: Decimal
    role: TransferPartyRole
    notes: str | None
    created_at: datetime


class OwnershipApprovalRecordResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    reviewer_user_id: UUID
    decision: OwnershipApprovalDecision
    comments: str | None
    created_at: datetime


class OwnershipTransferRequestResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    inventory_asset_id: UUID
    asset_display_id: str | None = None
    project_name: str | None = None
    transfer_type: TransferType
    effective_date: date
    reason: str
    supporting_document_id: UUID | None
    related_transaction_id: UUID | None
    requested_by_user_id: UUID
    assigned_approver_user_id: UUID | None
    status: TransferRequestStatus
    reviewed_at: datetime | None
    approved_at: datetime | None
    rejected_at: datetime | None
    decision_notes: str | None
    parties: list[TransferPartyResponse] = Field(default_factory=list)
    actions: list[str] = Field(default_factory=list)
    proposed_legal_total: Decimal | None = None
    current_legal_total: Decimal | None = None
    created_at: datetime
    updated_at: datetime


class OwnershipTransferRequestListResponse(BaseModel):
    items: list[OwnershipTransferRequestResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class OwnershipRecordResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    inventory_asset_id: UUID
    party_id: UUID
    party_name: str | None = None
    party_type: str | None = None
    ownership_type: OwnershipType
    ownership_percentage: Decimal
    effective_from: date
    effective_to: date | None
    status: OwnershipRecordStatus
    acquisition_method: AcquisitionMethod
    transfer_reason: str | None
    related_document_id: UUID | None
    related_transaction_id: UUID | None
    source: OwnershipSource
    notes: str | None
    created_by_user_id: UUID | None
    approved_by_user_id: UUID | None
    ownership_transfer_request_id: UUID | None
    created_at: datetime
    updated_at: datetime


class AssetOwnershipDetailResponse(BaseModel):
    asset_id: UUID
    current_legal: list[OwnershipRecordResponse]
    current_beneficial: list[OwnershipRecordResponse]
    current_economic: list[OwnershipRecordResponse]
    current_other: list[OwnershipRecordResponse]
    pending_requests: list[OwnershipTransferRequestResponse]
    scheduled_transfers: list[OwnershipTransferRequestResponse]
    history: list[OwnershipRecordResponse]
    legal_total: Decimal
    proposed_legal_total: Decimal | None = None


class OwnershipSummaryResponse(BaseModel):
    primary_legal_owner: str | None
    owner_count: int
    ownership_status: str


class OwnershipDecisionInput(BaseModel):
    comments: str | None = None
    decision_notes: str | None = None


class OwnershipPreviewResponse(BaseModel):
    current_legal_total: Decimal
    proposed_legal_total: Decimal
    valid: bool
    errors: list[str] = Field(default_factory=list)
    outgoing: list[TransferPartyResponse] = Field(default_factory=list)
    incoming: list[TransferPartyResponse] = Field(default_factory=list)
    continuing: list[TransferPartyResponse] = Field(default_factory=list)
